### DISCUSSION TITLE
Why do hooks in plugins loaded through the INI file not fire, but they do when loaded through `pytest.main()`?

### DISCUSSION THREAD

**Discussion by dgard1981** [pytest-dev/pytest#11210] [Q&A] [answered]
I have written a basic plugin that I wish to include via the INI file, but it doesn't seem to work. To demonstrate, take the example files below, which includes a plugin with two hooks -
- `pytest_assertion_pass` - Simply add 1 to a count of passed assertions.
- `pytest_terminal_summary` - Add "x assertions passed." to the end of the terminal summary.

It is possible to prove that the plugin is being loaded through the INI file because either -
- If the path is mis-typed then an import error occurs and `pytest` fails. or;
- If the plugin is include both in the INI file and through a `pytest.main()` call in code, a warning is displayed.
  > PytestAssertRewriteWarning: Module already imported so cannot be rewritten: plugins.count_assertions

Now using the example files with the plugin loaded through the INI file, notice how the plugin hooks are not fired (`x assertions passed.` is not displayed in the terminal summary).

```shell
user@computer:~/plugin-test$ python src
.                                                                                                                                                                                            [100%]
1 passed in 0.00s
user@computer:~/plugin-test$ pytest src -q
.                                                                                                                                                                                            [100%]
1 passed in 0.00s
```

Finally, to show that the plugin works, updating the `pytest.ini` and `__main__.py` like the example files with the plugin loaded through code, notice how the plugin hooks are fired as expected.

```shell
user@computer:~/plugin-test$ python src
.                                                                                                                                                                                            [100%]
2 assertions passed.
1 passed in 0.00s
```

So my question is, how can do you get plugins to work through the INI file? It's not practical to assume that a `pytest` run will be started through code every time, and one must assume that the CLI will be used more often than not.

---

### Example files - Plugin loaded through INI file

<details>
  <summary><code>src/pytest.ini</code></summary>

```
[pytest]
addopts = -p src.plugins.count_assertions
enable_assertion_pass_hook = true
```
</details>

<details>
  <summary><code>src/__main__.ini</code></summary>

```python
import pytest

args = ['src', '-q']
pytest.main(args)
```
</details>

<details>
  <summary><code>src/test_somthing.ini</code></summary>

```python
def test_stdout():
    assert 1 == 1
    assert 2 == 2
```
</details>

<details>
  <summary><code>src/plugins/count_assertions.py</code></summary>

```python
class PassedAssertions:

    count = 0

    def pytest_assertion_pass(self, *_) -> None:
        self.count += 1

    def pytest_terminal_summary(self, *_) -> dict:
        print(f'{self.count} assertions passed.')
```
</details>

---

### Example files - Plugin loaded through code

<details>
  <summary><code>src/pytest.ini</code></summary>

```
[pytest]
enable_assertion_pass_hook = true
```
</details>

<details>
  <summary><code>src/__main__.ini</code></summary>

```python
from src.plugins.count_assertions import PassedAssertions
import pytest

args = ['src', '-q']
pytest.main(args, plugins=[PassedAssertions()])
```
</details>

**Comment by dgard1981** [comment]
To add a little more information... if the plugin was updated as follows, `Module loaded` is printed regardless of how the plugin is included.

However, `Plugin initialised` is only printed if the plugin is loaded through code  - `pytest.main(args, plugins=[PassedAssertions()])`

So it seems that when including a plugin through the INI file, `pytest` doesn't look for and instantiate classes (which makes sense).

But manually instantiating the plugin class either in the plugin file or through `__init__.py` in the plugin directory also fails. Even though in both of those cases `Plugin initialised` is printed, presumably instantiation takes place too late.

```python
print('Module loaded')
class PassedAssertions:

    count = 0

    def __init__(self) -> None:
        print('Plugin initialised')

    def pytest_assertion_pass(self, *_) -> None:
        self.count += 1

    def pytest_terminal_summary(self, *_) -> dict:
        print(f'{self.count} assertions passed.')

```

**Reply by RonnyPfannschmidt** [reply]
You would need a pytest-configure hook to make and register a instance

**Reply by dgard1981** [reply]
Forgive my ignorance, but how would I do that?

Using the below, my plugin is instantiated, but the hooks within are still not fired.

```
def pytest_configure(*_) -> None:
    p = PassedAssertions()
```

**Reply by RonnyPfannschmidt** [reply, marked as answer]
Registration with the plugin manager missing

See the builtin junitxml plugin for a example

**Reply by dgard1981** [reply]
Thanks for the pointer.

Both of these seem to work, so no I'm just trying to find some reading on `config.stash` to see if it's relevant for my use case.

```python
def pytest_configure(config) -> None:
    config.stash['pytest_assertions_plugin'] = PytestAssertionsPlugin()
    config.pluginmanager.register(config.stash['pytest_assertions_plugin'])
```

```python
def pytest_configure(config) -> None:
    config.pluginmanager.register(PytestAssertionsPlugin())
```

**Reply by RonnyPfannschmidt** [reply]
The stash is for accessing objects You put there, if the plugin class is self contained it's not necessary
